db/src/lib.rs

```diff
@@ -0,0 +1,96 @@
+pub use sqlx;
+use sqlx::postgres::PgPoolOptions;
+pub use sqlx::PgPool;
+
+#[tracing::instrument(err)]
+pub async fn setup_db_pool() -> color_eyre::Result<PgPool> {
+    let database_url = std::env::var("DATABASE_URL").expect("DATABASE_URL must be set");
+    let pool = PgPoolOptions::new()
+        .max_connections(5)
+        .connect(&database_url)
+        .await?;
+    let mut connection = pool.acquire().await?;
+
+    let lock = sqlx::postgres::PgAdvisoryLock::new("knowledge-db-migration-lock");
+    let mut lock = lock.acquire(&mut connection).await?;
+
+    sqlx::migrate!().run(lock.as_mut()).await?;
+
+    lock.release_now().await?;
+    tracing::info!("Migration lock unlocked");
+
+    Ok(pool)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    mod test_db {
+        use sqlx::{migrate::MigrateDatabase, Postgres, Row};
+        use url::Url;
+
+        use super::*;
+        struct TestPool {
+            database_url: String,
+            pool: PgPool,
+        }
+
+        impl AsRef<PgPool> for TestPool {
+            fn as_ref(&self) -> &PgPool {
+                &self.pool
+            }
+        }
+
+        impl TestPool {
+            async fn teardown(self) -> color_eyre::Result<()> {
+                self.pool.close().await;
+                Postgres::drop_database(&self.database_url).await?;
+                Ok(())
+            }
+        }
+
+        async fn create_test_db_pool() -> color_eyre::Result<TestPool> {
+            let original_database_url = std::env::var("DATABASE_URL").unwrap();
+            let mut db_url = Url::parse(&original_database_url).unwrap();
+            let database_name = format!("/knowledge_test_{}", uuid::Uuid::new_v4());
+            db_url.set_path(&database_name);
+
+            let database_url = db_url.to_string();
+            Postgres::create_database(&database_url).await?;
+            std::env::set_var("DATABASE_URL", &database_url);
+
+            let pool = setup_db_pool().await?;
+
+            Ok(TestPool { database_url, pool })
+        }
+
+        #[tokio::test]
```

Comment:
You can use `#[tokio::test]` for tests that need an `async` function instead of a plan function.

I think there is a good change most of your tests end up being async so you'll want this version! Any function with the argument will automatically be run with you do `cargo test`